crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,341 @@
+                                        let result = transport
+                                            .lock()
+                                            .await
+                                            .send_op_request(
+                                                op_id.clone(),
+                                                agent_id.clone(),
+                                            )
+                                            .await;
+
+                                        // Request is complete, decrease request count.
+                                        (*current_request_count
+                                            .lock()
+                                            .await) -= 1;
+
+                                        if let Err(err) = result {
+                                            eprintln!("could not send fetch request for op id {op_id} to agent {agent_id}: {err}");
+                                            inner
+                                                .cool_down_list
+                                                .lock()
+                                                .await
+                                                .insert(
+                                                    agent_id,
+                                                    Instant::now(),
+                                                );
+                                        }
+                                    }
+                                });
+                            }
+
+                            // Pause to give other requests a chance to be processed before re-adding this op.
+                            tokio::time::sleep(Duration::from_millis(1)).await;
+
+                            // Push op and source back to the end of the queue.
+                            inner
+                                .ops
+                                .lock()
+                                .await
+                                .insert((op_id, agent_id), ());
+                        }
+                    }
+
+                    // Pause if maximum parallel request count is reached.
+                    if (*current_request_count.lock().await)
+                        == parallel_request_count
+                    {
+                        tokio::time::sleep(Duration::from_millis(
+                            parallel_request_pause,
+                        ))
+                        .await;
+                    } else if inner.ops.lock().await.is_empty() {
+                        // Sleep if there are no more ops to fetch.
+                        // Sleep is canceled by newly added ops.
+                        select! {
+                            _ = tokio::time::sleep(Duration::from_millis(
+                                fetch_loop_sleep,
+                            )) => {},
+                            _ = fetch_request_rx.recv() => {}
```

Comment:
Is this wake-up mechanism a premature optimization from your POV?

Reply:
I think the sleeping itself is re-inventing what channels are for. When you await on a channel, if there is no data, all the logic for pausing the task and re-waking it is built in.